### PR TITLE
[IAM]: ``resource_identity_role_v3`` refactoring + ``resource`` parameter added 

### PR DESCRIPTION
**Pull request by artem-lifshits**
## Summary of the Pull Request
Refactoring of ``resource_identity_role_v3`` including test update. New parameter ``resource`` added for ``statements`` field.

## PR Checklist

* [x] Refers to: #1767 
* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.
* [x] Release notes added.

## Acceptance Steps Performed

```
=== RUN   TestAccIdentityV3Role_importBasic
--- PASS: TestAccIdentityV3Role_importBasic (36.45s)
PASS

Process finished with the exit code 0


=== RUN   TestAccIdentityRoleV3_basic
--- PASS: TestAccIdentityRoleV3_basic (84.67s)
PASS

Process finished with the exit code 0

```

### CODE REVIEW

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go [hunk 1]
```diff
@@ -80,551 +88,190 @@ func ResourceIdentityRoleV3() *schema.Resource {
...
-	}
-	e, err = common.IsEmptyValue(reflect.ValueOf(typeProp))
-	if err != nil {
-		return nil, err
-	}
-	if !e {
-		params["type"] = typeProp
+	if d.HasChange("statement") {
+		opStatement := resourceIAMRoleV3OpStatement(d)
+		opts.Policy.Statement = opStatement
```
Comment (lego963) [inline opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: needsUpdate = true

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go [hunk 2]
```diff
@@ -80,551 +88,190 @@ func ResourceIdentityRoleV3() *schema.Resource {
...
-	if r.Err != nil {
-		return nil, fmt.Errorf("error run api(update): %s", r.Err)
+		return fmterr.Errorf("error creating identity v3.0 client: %s", err)
 	}
-	return r.Body, nil
-}
 
-func buildIdentityRoleV3UpdateParameters(opts map[string]interface{}, arrayIndex map[string]int) (interface{}, error) {
-	params := make(map[string]interface{})
+	opts := policies.CreateOpts{}
```
Comment (lego963) [inline opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: CreateOpts? maybe UpdateOpts?

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go [hunk 3]
```diff
@@ -80,551 +88,190 @@ func ResourceIdentityRoleV3() *schema.Resource {
...
-	}
-	return r.Body, nil
-}
-
-func buildIdentityRoleV3CreateParameters(opts map[string]interface{}, arrayIndex map[string]int) (interface{}, error) {
-	params := make(map[string]interface{})
-
-	descriptionProp, err := common.NavigateValue(opts, []string{"description"}, arrayIndex)
+	roleType, fmtErr := buildRoleType(d)
 	if err != nil {
```
Comment (lego963) [inline opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: `If fmtErr` should be

**Review by lego963 — CHANGES_REQUESTED** [review]

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go (hunk 1 above)
Comment (artem-lifshits) [reply opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: fixed

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go (hunk 2 above)
Comment (artem-lifshits) [reply opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: Update opts are same as CreateOpts

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go (hunk 3 above)
Comment (artem-lifshits) [reply opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: fixed

**Review by anton-sidelnikov — DISMISSED** [review]

**Review by vladimirvshivkov — DISMISSED** [review]

**Comment by artem-lifshits** [comment]
=== RUN   TestAccIdentityRoleV3_basic
--- PASS: TestAccIdentityRoleV3_basic (77.48s)
=== RUN   TestAccIdentityV3Role_importBasic
--- PASS: TestAccIdentityV3Role_importBasic (38.09s)
PASS

Process finished with the exit code 0

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go [hunk 4]
```diff
@@ -80,551 +93,178 @@ func ResourceIdentityRoleV3() *schema.Resource {
...
+	config := meta.(*cfg.Config)
+	client, err := common.ClientFromCtx(ctx, keyClientV30, func() (*golangsdk.ServiceClient, error) {
+		return config.IdentityV30Client()
+	})
 	if err != nil {
-		return fmt.Errorf("error reading Role:display_layer, err: %s", err)
-	}
-	if err = d.Set("display_layer", displayLayerProp); err != nil {
-		return fmt.Errorf("error setting Role:display_layer, err: %s", err)
+		return fmterr.Errorf("error creating identity v3.0 client: %s", err)
```
Comment (lego963) [inline opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: use const

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go [hunk 5]
```diff
@@ -80,551 +93,178 @@ func ResourceIdentityRoleV3() *schema.Resource {
...
-	req["Version"] = "1.1"
-
-	return req, nil
-}
-
-func expandIdentityRoleV3UpdatePolicyStatement(d interface{}, arrayIndex map[string]int) (interface{}, error) {
-	newArrayIndex := make(map[string]int)
-	for k, v := range arrayIndex {
-		newArrayIndex[k] = v
+		return fmterr.Errorf("error creating identity v3.0 client: %s", err)
```
Comment (lego963) [inline opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: same

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go [hunk 6]
```diff
@@ -80,551 +93,178 @@ func ResourceIdentityRoleV3() *schema.Resource {
...
-	newArrayIndex := make(map[string]int)
-	for k, v := range arrayIndex {
-		newArrayIndex[k] = v
+func resourceIdentityRoleV3Delete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*cfg.Config)
+	client, err := common.ClientFromCtx(ctx, keyClientV30, func() (*golangsdk.ServiceClient, error) {
+		return config.IdentityV30Client()
+	})
+	if err != nil {
+		return fmterr.Errorf("error creating identity v3.0 client: %s", err)
```
Comment (lego963) [inline opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: and here

**Review by lego963 — CHANGES_REQUESTED** [review]

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go (hunk 4 above)
Comment (artem-lifshits) [reply opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: thanks, fixed

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go (hunk 5 above)
Comment (artem-lifshits) [reply opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: fixed

File: opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go (hunk 6 above)
Comment (artem-lifshits) [reply opentelekomcloud/services/iam/resource_opentelekomcloud_identity_role_v3.go]: fixed

**Review by anton-sidelnikov — APPROVED** [review]

**Comment by artem-lifshits** [comment]
=== RUN   TestAccIdentityRoleV3_basic
--- PASS: TestAccIdentityRoleV3_basic (79.68s)
=== RUN   TestAccIdentityV3Role_importBasic
--- PASS: TestAccIdentityV3Role_importBasic (36.63s)
PASS

Process finished with the exit code 0

**Review by lego963 — APPROVED** [review]

**Pull request merged**